subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/ProgressLoggingFixture.groovy

```diff
@@ -89,6 +89,10 @@ class ProgressLoggingFixture extends InitScriptExecuterFixture {
                 private LoggingOutput loggingOutput
 
                 OutputProgressService() {
+                    // Ensure parameters.outputFile is set before adding ourselves as a listener
+                    if (!parameters.outputFile.present) {
```

Comment:
Where can I find the code that backs `BuildService.getParameters()`?

Reply:
Ah, I am looking now at https://github.com/gradle/gradle/blob/8e3faae63329c2ccdfa52d3a54d395fba3b03f52/subprojects/core/src/main/java/org/gradle/api/services/internal/BuildServiceProvider.java#L149

Reply:
So parameters are obtained via a service lookup ([this case](https://github.com/gradle/gradle/blob/e6d036e07cc8c358c31e07f3f23c402f590611df/subprojects/model-core/src/main/java/org/gradle/internal/isolated/IsolationScheme.java#L151)) that is available to a BuildService implementation (injected via [ASM](https://github.com/gradle/gradle/blob/d70090545b34a65edaf90843cd81fbaec9d152ac/subprojects/model-core/src/main/java/org/gradle/internal/instantiation/generator/AsmBackedClassGenerator.java#L1571))

Reply:
@bamboo Isn't the `getParameters()` accessor itself that is triggering the NPE? (because apparently there is no ObjectCreationDetails instance in the SERVICES_FOR_NEXT_OBJECT thread local)

I am not sure how that could happen, by at least for `getDisplayName()` we do consider the possibility that thread local may be empty (apparently during object construction? See this [PR](https://github.com/gradle/gradle/commit/e8e8d97d9ad65de4666602fb61d6b7cec35552ef#diff-a135921792daf52c5d8258172d93c9c675260977ce57409ff39fba4513f1a975R136)).

Reply:
`getParameters()` is (indirectly) triggering the NPE when it tries to use the injected services:
```
at org.gradle.internal.instantiation.generator.AsmBackedClassGenerator.getServicesForNext(AsmBackedClassGenerator.java:130)
at OutputProgressService$Inject.$gradleServices(Unknown Source)
at OutputProgressService$Inject.getParameters(Unknown Source)
```

> I am not sure how that could happen...

My hypothesis is that the service adds itself as a listener before being fully constructed then it gets a notification from a separate thread for which `SERVICES_FOR_NEXT_OBJECT` is not set.

By invoking `getParameters()` in the constructor before registering the service as a listener we're forcing Gradle to initialize the backing field before any notifications are delivered.

Reply:
Right, but then this would not help detecting that, right?

```
if (!parameters.outputFile.present) {
```

(as the access to parameters is the source of the NPE, not access to outputFile)

Reply:
That expression is there just to force the `parameters` property backing field to be initialized ahead of time, it would usually be initialized by the generated subclass constructor which runs **after** the user provided constructor.

Reply:
ohhh...